modules/axtask/src/run_queue.rs

```diff
@@ -75,54 +235,117 @@ impl AxRunQueue {
-        assert!(curr.is_running());
+        assert!(curr.is_running(), "task is not running: {:?}", curr.state());
         assert!(!curr.is_idle());
         if curr.is_init() {
-            EXITED_TASKS.lock().clear();
+            EXITED_TASKS.with_current(|exited_tasks| exited_tasks.clear());
             axhal::misc::terminate();
         } else {
             curr.set_state(TaskState::Exited);
-            curr.notify_exit(exit_code, self);
-            EXITED_TASKS.lock().push_back(curr.clone());
-            WAIT_FOR_EXIT.notify_one_locked(false, self);
-            self.resched(false);
+
+            // Notify the joiner task.
+            curr.notify_exit(exit_code);
+
+            // Push current task to the `EXITED_TASKS` list, which will be consumed by the GC task.
+            EXITED_TASKS.with_current(|exited_tasks| exited_tasks.push_back(curr.clone()));
+            // Wake up the GC task to drop the exited tasks.
+            WAIT_FOR_EXIT.with_current(|wq| wq.notify_one(false));
+            // Schedule to next task.
+            self.inner.resched(false);
         }
         unreachable!("task exited!");
     }
 
-    pub fn block_current<F>(&mut self, wait_queue_push: F)
-    where
-        F: FnOnce(AxTaskRef),
-    {
+    /// Block the current task, put current task into the wait queue and reschedule.
+    /// Mark the state of current task as `Blocked`, set the `in_wait_queue` flag as true.
+    /// Note:
+    ///     1. The caller must hold the lock of the wait queue.
+    ///     2. The caller must ensure that the current task is in the running state.
+    ///     3. The caller must ensure that the current task is not the idle task.
+    ///     4. The lock of the wait queue will be released explicitly after current task is pushed into it.
+    pub fn blocked_resched(&mut self, mut wq_guard: WaitQueueGuard) {
         let curr = crate::current();
-        debug!("task block: {}", curr.id_name());
         assert!(curr.is_running());
         assert!(!curr.is_idle());
-
         // we must not block current task with preemption disabled.
+        // Current expected preempt count is 2.
+        // 1 for `NoPreemptIrqSave`, 1 for wait queue's `SpinNoIrq`.
         #[cfg(feature = "preempt")]
-        assert!(curr.can_preempt(1));
+        assert!(curr.can_preempt(2));
 
+        // Mark the task as blocked, this has to be done before adding it to the wait queue
+        // while holding the lock of the wait queue.
         curr.set_state(TaskState::Blocked);
-        wait_queue_push(curr.clone());
-        self.resched(false);
+        curr.set_in_wait_queue(true);
+
+        wq_guard.push_back(curr.clone());
+        // Drop the lock of wait queue explictly.
+        drop(wq_guard);
+
+        // Current task's state has been changed to `Blocked` and added to the wait queue.
+        // Note that the state may have been set as `Ready` in `unblock_task()`,
+        // see `unblock_task()` for details.
+        assert!(
```

Comment:
This assertion seems obvious and can be deleted, will it be in the runnable or exit state?

Reply:
Nope, since it is `blocked_resched` and it accepts a `WaitQueueGuard` as a parameter, it will set current task's state as `Blocked` with holding the `WaitQueueGuard`, insert current task into wq and then drop the `WaitQueueGuard`.

It's what happens after the dropping of `WaitQueueGuard` that matters, 

since then current task maybe unblocked by other cores and set its state as `Ready` due to `unblock_task()`'s `task.transition_state(TaskState::Blocked, TaskState::Ready)`

Reply:
`assert(is_blocked || is_ready)` == `assert(is_not_running && is_not_exited)`，which is obvious.

BTW, `|` -> `||`

Reply:
I'll just delete the assertion but leave the comments